ethers-contract/ethers-contract-abigen/src/contract.rs

```diff
@@ -322,6 +321,101 @@ impl Context {
     pub fn internal_structs_mut(&mut self) -> &mut InternalStructs {
         &mut self.internal_structs
     }
+
+    /// Expands `self.extra_derives` into a comma separated list to be inserted in a
+    /// `#[derive(...)]` attribute.
+    pub(crate) fn expand_extra_derives(&self) -> TokenStream {
+        let extra_derives = &self.extra_derives;
+        quote!(#( #extra_derives, )*)
+    }
+
+    /// Generates the token stream for the contract's ABI, bytecode and struct declarations.
+    pub(crate) fn struct_declaration(&self) -> TokenStream {
+        let name = &self.contract_ident;
+
+        let ethers_core = ethers_core_crate();
+        let ethers_contract = ethers_contract_crate();
+
+        let abi = {
+            let abi_name = self.inline_abi_ident();
+            let abi = &self.abi_str;
+            let (doc_str, parse) = if self.human_readable {
+                // Human readable: use abi::parse_abi_str
+                let doc_str = "The parsed human-readable ABI of the contract.";
+                let parse = quote!(#ethers_core::abi::parse_abi_str(__ABI));
+                (doc_str, parse)
+            } else {
+                // JSON ABI: use serde_json::from_str
+                let doc_str = "The parsed JSON ABI of the contract.";
+                let parse = quote!(#ethers_core::utils::__serde_json::from_str(__ABI));
+                (doc_str, parse)
+            };
+
+            quote! {
+                #[rustfmt::skip]
+                const __ABI: &str = #abi;
+
+                // This never fails as we are parsing the ABI in this macro
+                #[doc = #doc_str]
+                pub static #abi_name: #ethers_contract::Lazy<#ethers_core::abi::Abi> =
+                    #ethers_contract::Lazy::new(|| #parse.expect("ABI is always valid"));
+            }
+        };
+
+        let bytecode = self.contract_bytecode.as_ref().map(|bytecode| {
+        let bytecode = bytecode.iter().copied().map(Literal::u8_unsuffixed);
+        let bytecode_name = self.inline_bytecode_ident();
+        quote! {
+            #[rustfmt::skip]
+            const __BYTECODE: &[u8] = &[ #( #bytecode ),* ];
+
+            #[doc = "The bytecode of the contract."]
+            pub static #bytecode_name: #ethers_core::types::Bytes = #ethers_core::types::Bytes::from_static(__BYTECODE);
+        }
+    });
+
+        quote! {
+            // The `Lazy` ABI
+            #abi
+
+            // The static Bytecode, if present
+            #bytecode
+
+            // Struct declaration
+            pub struct #name<M>(#ethers_contract::Contract<M>);
```

Comment:
we'll need to resolve conflict with https://github.com/gakonst/ethers-rs/commit/9733b16afc5c35be4409b72c172d8f7405f45d7c